### ISSUE TITLE
Unifying init() and setup() function into one

### ISSUE THREAD

**Issue by kristijanhusak** [creativenull/diagnosticls-configs-nvim#26]
Hi,

I would like to figure out why are `init()` and `setup()` functions separate? I'm thinking that simplifying it and using only a `setup()` function would be more straightforward. Something like this:

```lua
require 'diagnosticls-configs'.setup({
  default_config = false,
  format = false,
  on_attach = on_attach,
  filetypes = {}, -- same what current setup() accepts
})
```

I was creating a project specific setup and ran into this issue where I need to call both `init()` and `setup()` to override the functionality. In my `init.vim` I have `default_config = true`, but for one of my projects, I need to have specific settings, which I had to override with an `init({ default_config = false })` call. It wasn't fully clear to me at first that I need to do that.
I'm assuming this is done for a reason, but I can't figure out which one.

Generally, plugin works great, and thank you for building it. My diagnosticls setup started to grow over time and this cleaned it up tremendously.

**Comment by creativenull** [comment]
Hey glad you like it :)

For your use-case, not sure why it works that way but I do believe you don't need to call `init()` twice to override `default_config`, I believe you can still pass your custom linter/formatter config to `setup()` and it will prioritize that over `default_config`. Can you verify this with `:lua print(vim.inspect(vim.lsp.get_active_clients()))` and let me know if that works cuz I know I have it implemented but didn't get to test it.

As for the reason of two functions, it was to simplify how you would setup the linters/formatters for projects without the need to call the default `lspconfig.diagnosticls.setup()` for every project, separating LSP settings and linter/formatter settings.

`init()` was exclusively for custom LSP settings, for example - a custom `on_attach` handler, and for plugin settings.

`setup()` focuses more on configuring linter/formatter settings. Usually this will be called within your nvim config or it could be called [separately if you have a `.nvimrc` in your project](https://github.com/neovim/nvim-lspconfig/wiki/Project-local-settings). But I wanted to make sure that it was independent and free of any other dependency (besides diagnosticls-configs module).

In terms of unifying it, ATM I don't have any plans to, but I'm open to the idea. I'm guessing just using `setup()` and pass filetypes and LSP settings to it as two separate args could work.

**Comment by kristijanhusak** [comment]
I tested it the way you suggested, and it actually does not override the default config in project specific configuration.

So basically, I have this in my init.lua:
```lua
local dlsconfig = require('diagnosticls-configs')

dlsconfig.init({ default_config = true })
dlsconfig.setup()
```

And in my project specific `.nvimrc` that I manually source, I have this:

```vim
lua << EOF
local eslint = require('diagnosticls-configs.linters.eslint')
local prettier_eslint = require('diagnosticls-configs.formatters.prettier_eslint')
local dlsconfig = require('diagnosticls-configs')

dlsconfig.init({ default_config = false })
dlsconfig.setup({
  javascript = {
    linter = vim.tbl_extend('force', eslint, {
      args = {
        '--no-eslintrc',
        '-c',
        '.eslintrc.js',
        '--stdin',
        '--stdin-filename',
        '%filepath',
        '--format',
        'json'
      },
      requiredFiles = { '.eslintrc.js' },
      rootPatterns = { '.git' }
    }),
    formatter = prettier_eslint,
  }
})
EOF
```

If I remove `dlsconfig.init({ default_config = false })` from project configuration, custom `eslint` args are not picked up. It uses the default config always.

> In terms of unifying it, ATM I don't have any plans to, but I'm open to the idea. I'm guessing just using setup() and pass filetypes and LSP settings to it as two separate args could work.

That's exactly what I had in mind. I'm calling the setup anyway. But if you fix overriding the default config , I think that would be also fine, since there would be no need to call `init()` every time.

**Comment by creativenull** [comment]
Okay, so I just tested it out myself to see what I get and it kinda works?

From your config you don't need `setup()` in your init.lua, you want only one instance of `init()` and `setup()`. Try with commenting out `setup()` in your init.lua and commenting out `init()` in your .nvimrc

But there is one bug, if you manually change the linter with `vim.tbl_extend`, it doesn't reflect those changes with `default_config` on. However, if I include a different `sourceName` in the tbl_extend, it will take that and assign the proper extended config.

Will have to investigate why it doesn't override with use of `vim.tbl_extend` and `default_config`

Here is the changes I added that seems to work, in my case:

`init.lua`
```lua
local dlsconfig = require('diagnosticls-configs')

dlsconfig.init({ default_config = true })
```

`.nvimrc`
```vim
lua << EOF
local eslint = require('diagnosticls-configs.linters.eslint')
local prettier_eslint = require('diagnosticls-configs.formatters.prettier_eslint')
local dlsconfig = require('diagnosticls-configs')

dlsconfig.setup({
  javascript = {
    linter = vim.tbl_extend('force', eslint, {
      sourceName = 'eslint_extended',
      args = {
        '--no-eslintrc',
        '-c',
        '.eslintrc.js',
        '--stdin',
        '--stdin-filename',
        '%filepath',
        '--format',
        'json'
      },
      requiredFiles = { '.eslintrc.js' },
      rootPatterns = { '.git' }
    }),
    formatter = prettier_eslint,
  }
})
EOF
```

**Comment by kristijanhusak** [comment]
Yeah that would work, but this means that I need to run `setup()` manually for each project, right?

My idea is to have like a default setup for projects where I don't need any project specific settings, and in case I need to override it, I just do it in project specific configuration.

I did similar thing when I set up linters and formatters myself. I had the default configuration in the init.lua, and I started the server with custom configuration from project config when needed.

**Comment by creativenull** [comment]
Yes, you will have to run `setup()` manually for each project. But I see what you mean.

Could you try with just adding back the empty `setup()` in your init.lua but keep the same .nvimrc with the code I had above? I tried with that and it seems like it works, keeping both default configs and custom configs. Looks like an issue with extending custom configuration not working properly with default configs.

I will explore on the idea of unifying it into one call, this weekend.

**Comment by creativenull** [comment]
So I've had a dumb moment, I've been using my own project local plugin to figure out the issue 🤦

I've disabled that, and enabled `set exrc`, created a `.nvimrc` in my project. Followed the code I posted before, and made couple changes to troubleshoot and it turns out `eslint` config is the only one that doesn't update!

I extended `prettier_eslint` and that seem to be updated properly. I tried other formatters and linters extended and that also seem to work as well. Also enabled `default_config` and extended those configs within `setup()` and lo and behold, same result.

So I'm not sure why `eslint` itself isn't extending 😕

**Comment by kristijanhusak** [comment]
Ok, seems like it has to do something with `sourceName`. When I added the `sourceName = 'eslint'` property to my override, it properly registered it.
```lua
dlsconfig.setup({
  javascript = {
    linter = vim.tbl_extend('force', eslint, {
      args = {
        '--no-eslintrc',
        '-c',
        '.eslintrc.js',
        '--stdin',
        '--stdin-filename',
        '%filepath',
        '--format',
        'json'
      },
      requiredFiles = { '.eslintrc.js' },
      rootPatterns = { '.git' },
      sourceName = 'eslint', -- added this
    }),
    formatter = prettier_eslint,
  }
})
```

Not sure why that made it work, but it did. I guess you can investigate from that and see if it can be maybe solved so it's not required, or if not, to document it.

**Comment by kristijanhusak** [comment]
In the meantime I noticed that project that was causing issues for me had some `.eslintrc` files that ignored all nested folders. Seems like the default configuration works perfectly fine, without the need to override anything. I'm not sure if the overriding is still an issue, but I don't have any more of those here. Sorry for bothering you. Feel free to close this PR when you decide on it.

**Comment by creativenull** [comment]
So looks like I figured out what the issue was with the `sourceName`. When you have `default_config` enabled, the same config is used for different filetypes. So for `javascript` filetype, the default config will use the non-extended `eslint` config, and sourceName is unique to this. Defining a new config with the same sourceName will not work as it was already defined by `default_config` and diagnosticls would register that first (if it's defined first).

So I've updated the docs to highlight this ([see Advanced Config](https://github.com/creativenull/diagnosticls-configs-nvim#advanced-configuration)), and make sure that `sourceName` is now a required field when defining your own custom config separate from the default configs. This should be working as intended so no change is required.

As for unifying the two functions, I'd probably just leave them as is, because I don't see the need as of now. Unless there is more requests for it or issues stemming from this particular structure, I will keep this one as closed.
